packages/core/src/storage.ts

```diff
@@ -136,28 +168,171 @@ export function describeError(e: StorageOperation): string {
   return log.join('\n    ');
 }
 
-type StorageOperation = Operation<StorageItem, 'typechange' | 'rename' | 'replace'>;
+interface StorageItemDetailed {
+  contract: string;
+  label: string;
+  src: string;
+  type: ParsedTypeDetailed;
+}
+
+interface ParsedTypeDetailed extends ParsedTypeId {
+  item: TypeItem<ParsedTypeDetailed>;
+  args?: ParsedTypeDetailed[];
+  rets?: ParsedTypeDetailed[];
+}
+
+type StorageOperation<T = StorageItemDetailed> = Operation<T, 'typechange' | 'rename' | 'replace'>;
 
 export function getStorageUpgradeErrors(original: StorageLayout, updated: StorageLayout): StorageOperation[] {
-  function matchStorageItem(o: StorageItem, u: StorageItem) {
-    const nameMatches = o.label === u.label;
-
-    // TODO: type matching should compare struct members, etc.
-    const typeMatches = original.types[o.type].label === updated.types[u.type].label;
-
-    if (typeMatches && nameMatches) {
-      return 'equal';
-    } else if (typeMatches) {
-      return 'rename';
-    } else if (nameMatches) {
-      return 'typechange';
+  const originalDetailed = getDetailedLayout(original);
+  const updatedDetailed = getDetailedLayout(updated);
+  return getStorageUpgradeErrorsGeneric(originalDetailed, updatedDetailed, { canGrow: true });
+}
+
+interface StorageField {
+  label: string;
+  type: ParsedTypeDetailed;
+}
+
+// This function is generic because it can compare top level storage layout as well as struct layout.
+function getStorageUpgradeErrorsGeneric<T extends StorageField>(
+  original: T[],
+  updated: T[],
+  { canGrow }: { canGrow: boolean },
+): StorageOperation<T>[] {
+  const ops = levenshtein(original, updated, (a, b) => matchStorageField(a, b, { canGrow: false }));
```

Comment:
Shouldn't canGrow be true for the last item in the storage? Though I'm not sure if that can open the door to other kind of problems, and it's best to be conservative.

Reply:
Yeah technically the last item in storage can grow, I think it's safe and I'd love to support it. The issue is that `levenshtein` calls the match function without any indication that an element is the last one in the layout. We would have to change the levenshtein function to do that and I decided to go with the simpler conservative choice.

If people run into this limitation we will reconsider!